htlcswitch/link_test.go

```diff
@@ -180,6 +180,290 @@ func createInterceptorFunc(prefix, receiver string, messages []expectedMessage,
+//   -----add---->
+//   -----sig----x
+//   <----sig-----
+//   -----rev----x
+func TestChannelLinkSigThenRev(t *testing.T) {
+	t.Parallel()
+
+	const chanAmt = btcutil.SatoshiPerBitcoin * 5
+	const chanReserve = btcutil.SatoshiPerBitcoin * 1
+	aliceLink, bobChannel, batchTicker, start, cleanUp, restore, err :=
+		newSingleLinkTestHarness(chanAmt, chanReserve)
+	if err != nil {
+		t.Fatalf("unable to create link: %v", err)
+	}
+	defer cleanUp()
+
+	if err := start(); err != nil {
+		t.Fatalf("unable to start test harness")
+	}
+	defer aliceLink.Stop()
+
+	alice := newPersistentLinkHarness(
+		t, aliceLink, batchTicker, restore,
+	)
+
+	var (
+		coreLink  = aliceLink.(*channelLink)
+		aliceMsgs = coreLink.cfg.Peer.(*mockPeer).sentMsgs
+	)
+
+	ctx := linkTestContext{
+		t:          t,
+		aliceLink:  aliceLink,
+		aliceMsgs:  aliceMsgs,
+		bobChannel: bobChannel,
+	}
+
+	bobHtlc1 := generateHtlc(t, coreLink, 0)
+
+	// <-----add-----
+	// Send an htlc from Bob to Alice.
+	ctx.sendHtlcBobToAlice(bobHtlc1)
+
+	aliceHtlc1, _ := generateHtlcAndInvoice(t, 0)
+
+	// ------add---->
+	ctx.sendHtlcAliceToBob(0, aliceHtlc1)
+	ctx.receiveHtlcAliceToBob()
+
+	// ------sig----x
+	// Trigger a commitsig from Alice->Bob.
+	select {
+	case batchTicker <- time.Now():
+	case <-time.After(5 * time.Second):
+		t.Fatalf("could not force commit sig")
+	}
+
+	var msg lnwire.Message
+	select {
+	case msg = <-aliceMsgs:
+	case <-time.After(15 * time.Second):
+		t.Fatalf("did not receive message")
+	}
+
+	comSig, ok := msg.(*lnwire.CommitSig)
+	if !ok {
+		t.Fatalf("expected CommitSig, got %T", msg)
+	}
+
+	if len(comSig.HtlcSigs) != 1 {
+		t.Fatalf("expected 1 htlc sig, got %d", len(comSig.HtlcSigs))
+	}
+
+	// <-----sig-----
+	ctx.sendCommitSigBobToAlice(1)
+
+	// ------rev----x
+	select {
+	case msg = <-aliceMsgs:
+	case <-time.After(15 * time.Second):
+		t.Fatalf("did not receive message")
+	}
+
+	_, ok = msg.(*lnwire.RevokeAndAck)
+	if !ok {
+		t.Fatalf("expected RevokeAndAck, got %T", msg)
+	}
+
+	cleanUp = alice.restart(false, true)
+	defer cleanUp()
+
+	ctx.aliceLink = alice.link
+	ctx.aliceMsgs = alice.msgs
+
+	// --reestablish->
+	select {
+	case msg = <-ctx.aliceMsgs:
+	case <-time.After(15 * time.Second):
+		t.Fatalf("did not receive message")
+	}
+
+	_, ok = msg.(*lnwire.ChannelReestablish)
+	if !ok {
+		t.Fatalf("expected ChannelReestablish, got %T", msg)
+	}
+
+	// <-reestablish--
+	bobReest, err := bobChannel.State().ChanSyncMsg()
+	if err != nil {
+		t.Fatalf("unable to retrieve chan sync msg: %v", err)
+	}
+	ctx.aliceLink.HandleChannelUpdate(bobReest)
+
+	// ------add---->
+	// We don't send this to Bob as his channel state already contains this,
+	// but this demontrates the correct behavior on Alice's side.
+	select {
+	case msg = <-ctx.aliceMsgs:
+	case <-time.After(15 * time.Second):
+		t.Fatalf("did not receive message")
+	}
+
+	_, ok = msg.(*lnwire.UpdateAddHTLC)
+	if !ok {
+		t.Fatalf("expected UpdateAddHTLC, got %T", msg)
+	}
+
+	// ------sig---->
```

Comment:
seems like the two tests could be combined into one, with subtests that just switches the order at the end here?

Reply:
`TestChannelLinkRevThenSig` tests:
```
//   <----sig-----
//   -----rev----x
//   -----sig----x
```

`TestChannelLinkSigThenRev` tests:
```
//   -----sig----x
//   <----sig-----
//   -----rev----x
```



The order is slightly different in the 2nd case. We can't test:
```
//   <----sig-----
//   -----sig----x
//   -----rev----x
```
because we can't control when `RevokeAndAck` is sent in the link, but we can for `CommitSig` with the force ticker.  So when Bob sends a signature, Alice will always reply with a revocation first, there's no way to hold off on that. So to get the desired ordering in the second case, Alice sends her sig (which Bob doesn't receive), Bob sends his sig, and then Alice replies with a revoke (which Bob doesn't receive). Since they're different in this way, I don't think they should go in subtests.